src/fs/metadata_log_operations/create_file.hh

```diff
@@ -0,0 +1,127 @@
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#pragma once
+
+#include "fs/metadata_log.hh"
+#include "fs/path.hh"
+
+namespace seastar::fs {
+
+class create_file_operation {
+    metadata_log& _metadata_log;
+    bool _is_directory;
+    sstring _entry_name;
+    file_permissions _perms;
+    inode_t _dir_inode;
+    inode_info::directory* _dir_info;
+    ondisk_create_inode_as_dir_entry_header _ondisk_entry;
+
+    create_file_operation(metadata_log& metadata_log) : _metadata_log(metadata_log) {}
+
+    future<inode_t> perform_1(sstring path, file_permissions perms, bool is_directory) {
+        _is_directory = is_directory;
+        if (is_directory) {
+            while (not path.empty() and path.back() == '/') {
+                path.erase(path.end() - 1, path.end());
+            }
+        }
+
+        _entry_name = last_component(path);
+        if (_entry_name.empty()) {
+            if (is_directory) {
+                return make_exception_future<inode_t>(std::runtime_error("Invalid path"));
+            } else {
+                return make_exception_future<inode_t>(std::runtime_error("Path has to end with character different than '/'"));
+            }
+        }
+
+        _perms = perms;
+        path.erase(path.end() - _entry_name.size(), path.end());
+        return _metadata_log.futurized_path_lookup(path).then([this](inode_t dir_inode) {
+            _dir_inode = dir_inode;
+            return with_semaphore(_metadata_log._create_or_delete_lock, 1, [this] {
+                // We do not have to shared lock dir_inode because we hold global lock
+                return perform_2();
+            });
+        });
+    }
+
+    future<inode_t> perform_2() {
+        auto dir_it = _metadata_log._inodes.find(_dir_inode);
+        if (dir_it == _metadata_log._inodes.end()) {
+            return make_exception_future<inode_t>(operation_became_invalid_exception());
+        }
+
+        assert(dir_it->second.is_directory());
+        _dir_info = &dir_it->second.get_directory();
+
+        return _metadata_log._dir_entry_locks.with_lock_on({_dir_inode, _entry_name}, [this] {
+            return perform_3();
+        });
+    }
+
+    future<inode_t> perform_3() {
+        if (_dir_info->entries.count(_entry_name) != 0) {
+            return make_exception_future<inode_t>(file_already_exists_exception());
+        }
+
+        decltype(_ondisk_entry.entry_name_length) entry_name_length;
+        if (_entry_name.size() > std::numeric_limits<decltype(entry_name_length)>::max()) {
+            return make_exception_future<inode_t>(filename_too_long_exception());
+        }
+        entry_name_length = _entry_name.size();
+
+        using namespace std::chrono;
+        uint64_t now_ns = duration_cast<nanoseconds>(system_clock::now().time_since_epoch()).count();
+        unix_metadata unx_mtdt = {
+            _perms,
+            getuid(), // TODO: this or something else?
+            getgid(), // TODO: this or something else?
```

Comment:
0 is also a fine placeholder for now, but getuid and getgid are natural candidates, so it's fine, although local uid/gid do not provide much context for a distributed system. Eventually, we'll want a user to be able to pass his credentials when bootstrapping the file system - that will allow us to authorize users on startup (e.g. via LDAP or whatnot). You can add a TODO comment about it around here.